### PR TITLE
add scrapers playground

### PR DESCRIPTION
**Pull request by esakal**
An inner project playground that can be used to maintain/extend scrapers. By using the playground:
- it simplify the workstation setup needed to run the library for development purposes.
- it makes IDE debugging tools available and useful

---
This change is [Reviewable](https://reviewable.io/reviews/eshaham/israeli-bank-scrapers/91)

### CODE REVIEW

File: playground/amex.tpl.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+const options = {
+  companyId: 'amex',
+  startDate: moment('2017-08-01'),
+  combineInstallments: false,
+  showBrowser: true,
+  verbose: true,
+};
+
+
```
Comment (eshaham) [inline playground/amex.tpl.js]: @esakal redundant line break

File: playground/discount.tpl.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+const options = {
+  companyId: 'discount',
+  startDate: moment('2017-08-01'),
+  combineInstallments: false,
+  showBrowser: true,
+  verbose: true,
+};
+
+
```
Comment (eshaham) [inline playground/discount.tpl.js]: @esakal redundant line break

File: playground/hapoalim.tpl.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+const options = {
+  companyId: 'hapoalim',
+  startDate: moment('2017-08-01'),
+  combineInstallments: false,
+  showBrowser: true,
+  verbose: true,
+};
+
+
```
Comment (eshaham) [inline playground/hapoalim.tpl.js]: @esakal redundant line break

File: playground/isracard.tpl.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+const options = {
+  companyId: 'isracard',
+  startDate: moment('2017-08-01'),
+  combineInstallments: false,
+  showBrowser: true,
+  verbose: true,
+};
+
+
```
Comment (eshaham) [inline playground/isracard.tpl.js]: @esakal redundant line break

File: playground/leumi-card.tpl.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+const options = {
+  companyId: 'leumiCard',
+  startDate: moment('2017-08-01'),
+  combineInstallments: false,
+  showBrowser: true,
+  verbose: true,
+};
+
+
```
Comment (eshaham) [inline playground/leumi-card.tpl.js]: @esakal redundant line break

File: playground/leumi.tpl.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+const options = {
+  companyId: 'leumi',
+  startDate: moment('2017-08-01'),
+  combineInstallments: false,
+  showBrowser: true,
+  verbose: true,
+};
+
+
```
Comment (eshaham) [inline playground/leumi.tpl.js]: @esakal redundant line break

File: playground/visa-cal.tpl.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+const options = {
+  companyId: 'visaCal',
+  startDate: moment('2017-08-01'),
+  combineInstallments: false,
+  showBrowser: true,
+  verbose: true,
+};
+
+
```
Comment (eshaham) [inline playground/visa-cal.tpl.js]: @esakal redundant line break

File: playground/helpers/utils.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
+import fs from 'fs';
+import path from 'path';
+import json2csv from 'json2csv';
+import moment from 'moment';
+
+const writeFileAsync = promisify(fs.writeFile);
+const makeDirAsync = promisify(fs.mkdir);
+const accessAsync = promisify(fs.access);
+
+export function promisify(fn) {
```
Comment (eshaham) [inline playground/helpers/utils.js]: @esakal this function already exists under `util` module, see [here](https://nodejs.org/dist/latest-v8.x/docs/api/util.html#util_util_promisify_original)

File: playground/helpers/utils.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    })
+  }
+}
+
+async function verifyFolder(filePath) {
+  const folder = path.dirname(filePath);
+
+  try {
+    await accessAsync(folder);
+  }catch(err)
```
Comment (eshaham) [inline playground/helpers/utils.js]: @esakal code style
```node
} catch (err) {
  // ...
}
```

File: playground/helpers/utils.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  {
+    await makeDirAsync(folder);
+  }
+}
+
+
```
Comment (eshaham) [inline playground/helpers/utils.js]: @esakal redundant line break

**Comment by esakal** [comment]
@eshaham I fixed the changes as requested but I didn't pushed them yet because I'm having a second thought about my implementation.

I think having a playground is important because it allows you to easily debug your code with common IDEs by setting breakpoints. That being said, I don't feel comfortable having a file with my personal credentials inside a folder that is version controlled. These might lead to undesired scenarios like pushing those files by mistake or deploying them to NPM (for example in my original PR commit I forgot to exclude folder `playground` from deployed NPM packages).

I think we should discard and close this PR without merging it.

But I still think that having a `safe playground` is important for this project, If you agree I can an improved minimal playground support with the following structure:
root
   | -- playground
      | -- setup.js
      | -- scrape.js

Where in `setup.js` you will be able to save encrypted credentials into the os home folder and with `scrape.js` you will be able to run them. we already have the `scrape.js` file I will just need to change it to support loading the credentials from the os home folder.

I don't mind creating a PR for that so you will be able to "feel" it and then decide if to merge it or not. 

what do you think?

Regarding this PR, we can close it once we complete this discussion.

**Comment by eshaham** [comment]
@esakal right

one of my comments for this PR was - let's change all `tpl` files to samples, instruct users to duplicate and remove samples ending and make sure to add that to `.gitignore`. See my comment [above](https://github.com/eshaham/israeli-bank-scrapers/pull/91#pullrequestreview-102873875).

Having said that, it is indeed worrisome that someone will end up publishing their credentials.
I like the approach you suggested. I just wonder if we could somehow share/borrow code from https://github.com/eshaham/israeli-ynab-updater, since it does similar stuff. wdyt?

**Comment by esakal** [comment]
Great, we can certainly borrow most of the code from israeli-ynab-updater. I will work on a new PR instead of this one. 

I think we can close this one

**Comment by eshaham** [comment]
@esakal 👍

**Comment by esakal** [comment]
@eshaham I'm closing this PR, I will create a new one based on our discussion once I complete my work on eshaham/israeli-ynab-updater#28

**Comment by eshaham** [comment]
@esakal :+1:

**Pull request closed without merging**